src/string_array.c

```diff
@@ -133,6 +134,47 @@ rcutils_string_array_cmp(
   return RCUTILS_RET_OK;
 }
 
+rcutils_ret_t
+rcutils_string_array_resize(
+  rcutils_string_array_t * string_array,
+  size_t new_size)
+{
+  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
+    string_array, "string_array is null", return RCUTILS_RET_INVALID_ARGUMENT);
+
+  if (string_array->size == new_size) {
+    return RCUTILS_RET_OK;
+  }
+
+  // Reclaim entries being removed
+  rcutils_allocator_t * allocator = &string_array->allocator;
+  if (!rcutils_allocator_is_valid(allocator)) {
+    RCUTILS_SET_ERROR_MSG("allocator is invalid");
+    return RCUTILS_RET_INVALID_ARGUMENT;
+  }
+  for (size_t i = new_size; i < string_array->size; ++i) {
+    allocator->deallocate(string_array->data[i], allocator->state);
+    string_array->data[i] = NULL;
+  }
+
+  char ** new_data = allocator->reallocate(
+    string_array->data, new_size * sizeof(char *), allocator->state);
+  if (NULL == new_data && 0 != new_size) {
+    RCUTILS_SET_ERROR_MSG("failed to allocate string array");
+    return RCUTILS_RET_BAD_ALLOC;
+  }
```

Comment:
Alright, I implemented this in adeb2a0, but it does make the function a lot more complicated than it was before, and introduces a possibility for a memory leak during error handling, which isn't great.

I opted to do it this way rather than always malloc/copy the data because I'm betting that "grow" operations will be more common than "shrink" ones, and realloc may be able to just extend the memory block, making this a very efficient operation.